demoapp/demo_app_functions.py

```diff
@@ -450,4 +466,479 @@ def update_csv(bill_num: str, title: str, summarized_bill: str, category: str, t
     
     df.to_csv(csv_file_path, index=False)
     return df
+
+class Container: 
+
+    '''
+    A container class to store all the details pertinant to that specific action
+    '''
+
+    def __init__(self, **kwargs):
+        for k, v in kwargs.items(): setattr(self, k, v)
+
+class BillDetails(Container): pass
+
+class LLMResults(Container): pass
+
+def get_llm_call_type(bill_details: BillDetails) -> str:
+
+    """
+    This function calculates number of tokens and decides on weather to use RAG or not. It reutrns a string output
+    that specifies how to call the LLM. 
+
+    Args: 
+        bill_details (BillDetails): object consisting of bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        str: 'large' or 'small' depeneding upon token count
+
+    """
+
+    bill_text = getattr(bill_details, "bill_text")
+    bill_title = getattr(bill_details, "bill_title")
+    mgl_ref = getattr(bill_details, "mgl_ref")
+    committee_info = getattr(bill_details, "committee_info")
+    mgl_names = getattr(bill_details, "mgl_names")
+
+    num_tokens = count_tokens(bill_title, bill_text, mgl_ref, mgl_names, committee_info)
+    
+    return 'small' if num_tokens < 120000 else 'large'
+
+def get_category_tags(categories: List) -> List:
+
+    """
+    This function takes in list of categories and returns tags pertinant to that specifc categories only. 
+
+    Args: 
+
+        categories (List(str)): List of category strings.
+
+    Returns: 
+        List of all tags specific to those of categories.  
+    """
+
+    tags_tuple = itemgetter(*categories)(new_tags_for_bill_dict)
+    category_tags = []
+    for cts in tags_tuple: category_tags += cts
+    return category_tags
+
+def generate_outputs(bill_details: BillDetails, task: str, return_metrics: bool = False) -> tuple:
+
+    '''
+    This function takes in bill details object (bill title, bill text and reference mgl section text) and completes
+    the task of either summarizing or tagging the bill. 
+
+    Arguments: 
+
+    bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+    task (str): It can take 2 values ('summarize','tag')
+    return_metrics (bool): Boolean values indication weather to calculate and return metrics or not - Defaults to False
+
+    Returns: 
+        A tuple of status_code and an LLMResults object containing query, response from the LLM and metrics - if requested. 
+
+        status_code can take these following values {1: Success, -1: Necessary details not found, -2: Invalid task}
+
+    '''
+
+    if not all(hasattr(bill_details, attr) for attr in ("bill_text", 'bill_title', 'mgl_names', 'committee_info')): 
+        return -1, LLMResults()
+
+    llm_call_type = get_llm_call_type(bill_details)   
+
+    if task == 'summarize': 
```

Comment:
Recommend defining two separate functions rather than using the conditional logic here.